extensions/amp-sidebar/1.0/sidebar.js

```diff
@@ -146,18 +147,49 @@ function SidebarWithRef(
   useLayoutEffect(() => {
     const sidebarElement = sidebarRef.current;
     const backdropElement = backdropRef.current;
-    if (!sidebarElement || !backdropElement) {
-      return;
-    }
-
-    if (!side) {
+    if (!side || !sidebarElement || !backdropElement) {
       return;
     }
 
     let sidebarAnimation;
     let backdropAnimation;
+    let keydownCallback;
+
+    const postInvisibleAnim = () => {
+      if (onAfterCloseRef.current) {
+        onAfterCloseRef.current();
+      }
+      sidebarAnimation = null;
+      backdropAnimation = null;
+      setMounted(false);
+    };
+    const documentElement =
+      sidebarElement.ownerDocument &&
+      sidebarElement.ownerDocument.documentElement;
+
     // "Make Visible" Animation
     if (opened) {
+      if (documentElement) {
+        keydownCallback = (event) => {
+          // Close sidebar on ESC.
+          if (event.key == Keys.ESCAPE) {
+            if (sidebarAnimation) {
+              sidebarAnimation.onfinish = () => {
+                postInvisibleAnim();
+                setOpened(false);
+              };
+              sidebarAnimation.reverse();
+            }
+            if (backdropAnimation) {
+              backdropAnimation.reverse();
+            }
+            event.preventDefault();
+            documentElement.removeEventListener('keydown', keydownCallback);
+          }
+        };
+        documentElement.addEventListener('keydown', keydownCallback);
```

Comment:
I'd recommend declaring `onKeyDown` prop on the sidebar element (`ContainWrapper`), much like `amp-lightbox` - you should be able to do something like: `onKeyDown={(event) => if (Esc) close()}`. 

The only issue is that when you first open the sidebar via the "open" button, focus stays on that DOM element - as long as we place focus inside the sidebar after the open sequence (perhaps in `postVisibleAnim` we can add `tryFocus(sidebarElement)`), then we should be able to capture key down events to the sidebar target rather than listening on the whole document. 

This brings up a good accessibility point - focus should be inside the sidebar anyway once it opens. And perhaps separately, we should confirm that when focus leaves the sidebar tree, it will also close. This is essentially the keyboard input version of clicking on the backdrop.

Reply:
That's a great idea, I like how this keeps it fully declarative as well.

There is one use case that I wanted to address that I think only works within effects, but please correct me if I am wrong.  So I wanted to be able to detect if the sidebar is currently in the open animation, and if so reverse that animation from where it is, instead of doing a call to `close()`.  This would make it so that the sidebar starts animating backwards from wherever it has opened to, whereas a call to `close()` automatically fully opens the sidebar and starts closing from the fully opened position.  It takes a little more code, but I think its a nice effect and keeps us in sync with `0.1`.

Regarding the `focus` I think you make some good points.  How do you feel about my addressing in a separate PR that is dedicated to `focus` and `a11y` management.  And then I can take a further look at where the `eventListener` is located then?

Reply:
If we can control focus, this is a good idea. But, also, to stress: a global document listener for things like this is a very common thing. So I'd not be offended if that's where we end up. It's kind of similar to history integration too where we basically have a global listener. As long as it's set and remove at the right time - it's good.

But that being said - it's also important to clarify the focus story with the sidebar.

Reply:
Separate PR for focus sounds good to me. Thanks for clarifying about document listeners as well!